vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,416 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace rec Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Windows.Controls
+open System.Windows.Media
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Formatting
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio.Utilities
+open Microsoft.CodeAnalysis
+open System.Threading
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open Microsoft.VisualStudio.Shell
+open Microsoft.VisualStudio
+open Microsoft.VisualStudio.LanguageServices
+open System.Windows
+open System.Collections.Generic
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.Ast
+open Microsoft.CodeAnalysis.Editor.Shared.Extensions
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Classification
+open Internal.Utilities.StructuredFormat
+open Microsoft.VisualStudio.Text.Tagging
+open System.Collections.Concurrent
+open System.Collections
+open System.Windows.Media.Animation
+open System.Globalization
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+open Microsoft.CodeAnalysis.Text
+open System
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text
+
+[<AbstractClass>]
+type CodeLensDisplayService (view : IWpfTextView, buffer : ITextBuffer) =
+
+    /// <summary>
+    /// Enqueing an unit signals to the tagger that all visible line lens must be layouted again,
+    /// to respect single line changes.
+    /// </summary>
+    member val RelayoutRequested : Queue = Queue() with get
+
+    member __.WpfView = view
+
+    member __.TextBuffer = buffer
+    
+    /// Public non-thread-safe method to add line lens for a given tracking span.
+    /// Returns an UIElement which can be used to add Ui elements and to remove the line lens later.
+    abstract AddCodeLens : ITrackingSpan -> UIElement
+    
+    /// Public non-thread-safe method to remove line lens for a given tracking span.
+    /// Returns whether the operation succeeded
+    abstract RemoveCodeLens : ITrackingSpan -> unit
+
+    abstract AddUiElementToCodeLens : ITrackingSpan -> UIElement -> unit
+    
+    abstract AddUiElementToCodeLensOnce : ITrackingSpan -> UIElement -> unit
+
+    abstract RemoveUiElementFromCodeLens : ITrackingSpan -> UIElement -> unit
+
+type CodeLensGeneralTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag:obj, providerTag:obj) =
+    inherit SpaceNegotiatingAdornmentTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag, providerTag)
+
+/// Class which provides support for general code lens
+/// Use the methods <code>AddCodeLens</code< and <code> RemoveCodeLens</code>
+type CodeLensGeneralTagger (view, buffer) as self =
```

Comment:
Consider to create a complete abstract class from  both implementations. I see way too much duplicate code.

Reply:
@dsyme in sense of OOP I would have used here a class which has protected members. Any design suggestions here?